PR: Switch to new API for Spyder 5

skjerns opened the pull request:
I've implemented the requested changes

* Now fully compatible with the new Spyder 5 Plugin system
* Version has `spyder>=5` as a hard requirement, no dual versions
* Seems to work as well as before 🥳

Closes #54 
Closes #53 
Closes #52

skjerns commented:
It's ready to review @ethanbb @dalthviz @jitseniesen :)

jitseniesen commented:
Great work, thanks! I will review in a couple of days.

dalthviz commented on spyder_line_profiler/tests/test_lineprofiler.py:
```diff
@@ -44,10 +43,10 @@ def test_profile_and_display_results(qtbot, tmpdir, monkeypatch):
-    monkeypatch.setattr('spyder_line_profiler.widgets.lineprofiler.QMessageBox',
-                        MockQMessageBox)
+    # monkeypatch.setattr('spyder_line_profiler.widgets.lineprofiler.QMessageBox',
+    #                     MockQMessageBox)
```
Should this be removed?

skjerns replied:
removed it

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -217,6 +236,25 @@ def __init__(self, parent):
+        #self.add_item_to_menu(
+        #    lambda x:x,
+        #    menu,
+        #)
```
Should this be removed?

skjerns replied:
removed it

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,113 @@
+        #     register_shortcut=True,
+        # )
+
+        # run_action.setEnabled(is_profiler_installed())
```
To add the action to the corresponding menu I think you could do something like:

```python
self._main.run_menu_actions += [lineprofiler_act]
```
In the new API the plugins have a reference to the main window through `self._main` instead of `self.main`

skjerns replied:
I don't think we need this, or what is the purpose of this?

I would not like to add a button to the menu. I don't think it's necessary, and the panel is enough

dalthviz replied:
I think for the moment preserving the way the plugin was working is better and also seems that some people use it so I think is worthy to keep it as for the issues here: https://github.com/spyder-ide/spyder-line-profiler/issues/57

However, maybe later we can check the removal of the action from the menu.

To create the entry in the menu then we can remove the lines here and instead add a method to handle the connection with the main menu as is done in the Profiler plugin: https://github.com/spyder-ide/spyder/blob/e948501d201d77c1dff092bdc784310211f452d0/spyder/plugins/pylint/plugin.py#L126-L132

dalthviz commented on setup.py:
```diff
@@ -38,13 +38,13 @@ def get_package_data(name, extlist):
 
 # Requirements
-REQUIREMENTS = ['line_profiler', 'spyder>=4']
+REQUIREMENTS = ['line_profiler', 'spyder>=5']
```
Not sure if the constraint should be `>=5.2` instead of `>=5`

skjerns replied:
that's true, it should be 5.2

dalthviz commented on spyder_line_profiler/__init__.py:
```diff
@@ -1,14 +1,14 @@
 # -*- coding: utf-8 -*-
+# ----------------------------------------------------------------------------
+# Copyright © 2022, Spyder Line Profiler 5
```
Maybe, for the files preserved in the API switch, this could be something like:
``` python
# Copyright © 2013 - Spyder Line Profiler contributors
```

 and,  for the new files, something like:

``` python
# Copyright © 2022 - Spyder Line Profiler contributors
```

?

skjerns replied:
I left the copyright everywhere the same and replaced the new ones with 2022. I think almost no file stayed the same.

dalthviz reviewed (commented):
Hi @skjerns thank you so much for your help and patience with this one! I left some initial comments

dalthviz commented on spyder_line_profiler/spyder/confpage.py:
```diff
@@ -0,0 +1,58 @@
+
+        # Warning: do not try to regroup the following QLabel contents with
+        # widgets above -- this string was isolated here in a single QLabel
+        # on purpose: to fix Issue 863 of Profiler plugon
```
```suggestion
        # on purpose: to fix Issue 863 of Profiler plugin
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -138,13 +153,13 @@ def __init__(self, parent):
         self.filecombo = PythonModulesComboBox(self)
 
         self.start_button = create_toolbutton(
```
The actions creation should be in the `setup` method. I think a good example of the structure this should follow is the Profiler plugin: https://github.com/spyder-ide/spyder/blob/master/spyder/plugins/profiler/widgets/main_widget.py

Also, instead of creating the actions as tool buttons, lets create them as actual actions to add them shortcuts so for example the `start_button` should become `start_action` and be create with something like:

```python
self.start_action = self.create_action(
            LineProfilerWidgetActions.Run,
            text=_("Profile by line"),
            tip=_("Run line profiler"),
            icon=self.create_icon('run'),
            triggered=self.analyze,
        )
```

Where the `LineProfilerWidgetActions` works as a holder for constants to identify the actions and then all the actions create can be added to the default main toolbar that the new API provides with something like the for as is used in the Profile plugin:

```python
 toolbar = self.get_main_toolbar()
        for item in [self.filecombo, browse_action, self.start_action, self.stop_action]:
            self.add_item_to_toolbar(
                item,
                toolbar=toolbar,
                section=LineProfilerWidgetMainToolbarSections.Main,
            )

```

Where `LineProfilerWidgetMainToolbarSections` defines constants for the sections available for the main toolbar.

Also, for the icons please use the `self.create_icon` method (in that way the plugin can use the new set of icons that the rest of Spyder is currently using). I think the names of the icons to use are for the main toolbar 'fileopen', 'start', 'stop', and for the secondary toolbar `collapse`, `expand` and 'log'.

In that way the layout of the plugin will be more similar to the rest of the migrated plugins

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,113 @@
+
+        # run_action.setEnabled(is_profiler_installed())
+        # editor = self.get_plugin(Plugins.Editor)
+        # widget.sig_edit_goto_requested.connect(editor.load)
```
I think we can remove this lines since they are already covered with the `on_editor_available` method

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -68,7 +56,27 @@
+    _ = get_translation("line_profiler", dirname="spyder_line_profiler")
+except KeyError as error:
+    import gettext
+    _ = gettext.gettext
```
Could you explain a little bit more why this is needed please? Is this related with the manually added `.pot` file?

skjerns replied:
No idea, I copied this line from the previous version. As the `.pot` file is empty anyway, we can probably remove it.

dalthviz replied:
Oh I see, maybe this lines are not needed since the translation function assigment is covered already by https://github.com/spyder-ide/spyder-line-profiler/blob/3ba3cf53538cf617c11a4ad2037be1aff562dd10/spyder_line_profiler/spyder/widgets.py#L44  ?

dalthviz reviewed (commented):
Thank you @skjerns for taking into account the initial review! Left some more comments regarding the toolbuttons creation, the entry in the run menu and the handling of the translations after checking locally the plugin. Functionally seems like is working :+1:

![imagen](https://user-images.githubusercontent.com/16781833/158875033-23f6141b-cf46-4060-be1a-a613325a9f3f.png)

skjerns commented:
@dalthviz thanks for the comments!

I've implemented everything as you wished :)

Additionally, as I was already at it, I added some more functionality

- add an elapsed timer
- added button to save results

In total this should
fix #48
fix #27 
fix #54 
fix #52

skjerns commented:
So, finally fixed all the tests.

If youn let me where I can find a full list of valid spyder icon names and how they look like, let me know. Then I'll use those.

skjerns commented:
Found a workaround :) patched the given functions to also work with previous versions of spyder.

All done & ready for review!

dalthviz commented:
Thank you so much for all the work here @skjerns !

> If you let me where I can find a full list of valid spyder icon names and how they look like, let me know. Then I'll use those.

You can find the available icons at https://github.com/spyder-ide/spyder/blob/8716a713a5060f40e5ce9cdc67b183829522a37d/spyder/utils/icon_manager.py#L111 I will make some suggestion in the icons lines with some screenshots in the next review :+1: 

> By using the more up-to-date plugin API we rely on changes only available after 5.1.5.
> 
> Therefore I think we need to have a hard limit for Spyder>=5.1.5, however that would probably cause problems when using older Python versions (see CI failurs.). @dalthviz how should we proceed?
> 
> 

Regarding the Spyder version on the CI I think we should config the workflows/jobs to use the conda-forge channel (since the Anaconda default channel goes only up to Spyder 5.1.5) in that way I think we should be able to not have to monkey patch or workaround things for the CI to pass. I think the `setup-miniconda` github action has a config to select the channel priority. I will let you a suggestion in the corresponding lines too :+1:

dalthviz commented on .github/workflows/macos-tests.yml:
```diff
@@ -19,12 +19,12 @@ jobs:
+        uses: conda-incubator/setup-miniconda@v2
         with:
            activate-environment: test
            auto-update-conda: true
```
We need to add here the config for the channel priority. It should be something like:

```suggestion
           auto-update-conda: true
           channels: conda-forge
           channel-priority: strict
```

dalthviz commented on .github/workflows/windows-tests.yml:
```diff
@@ -19,12 +19,12 @@ jobs:
+        uses: conda-incubator/setup-miniconda@v2
         with:
            activate-environment: test
            auto-update-conda: true
```
We need to add here the config for the channel priority. It should be something like:
```suggestion
           auto-update-conda: true
           channels: conda-forge
           channel-priority: strict
```

dalthviz commented on .github/workflows/linux-tests.yml:
```diff
@@ -28,7 +28,7 @@ jobs:
+        uses: conda-incubator/setup-miniconda@v2
         with:
            activate-environment: test
            auto-update-conda: true
```
We need to add here the config for the channel priority. It should be something like:
```suggestion
           auto-update-conda: true
           channels: conda-forge
           channel-priority: strict
```

dalthviz commented on conda.recipe/meta.yaml:
```diff
@@ -24,7 +24,7 @@ requirements:
     - line_profiler
     - python
-    - spyder >=3
+    - spyder >=5
```
We should put 5.2 here
```suggestion
    - spyder >=5.2
```

dalthviz commented on requirements/conda.txt:
```diff
@@ -1,2 +1,2 @@
 line_profiler
-spyder >=4.0.0b5
+spyder >=5
```
We should put 5.2 here:
```suggestion
spyder >=5.2
```

dalthviz commented on spyder_line_profiler/__init__.py:
```diff
@@ -1,14 +1,14 @@
+Spyder Line Profiler.
+"""
+
+__version__ = "0.5.1.dev0"
```
I think the next release will be 0.3.0 so this should still be 0.3.0.dev0 

However, maybe I missing something, is there some reason to jump to 0.5?

skjerns replied:
I thought it might make sense to reflect the compatibility jump from Spyder 3 to Spyder 5, by having the minor match the python major. At least I'd suggest that. But it doesn't matter much.

dalthviz replied:
Oh I see I think using 0.3.0 should be enought although maybe following the idea of a way to reflect the compatibility jump we should directly jump to 1.0.0? What do you think @ccordoba12 @jitseniesen ?

dalthviz replied:
@skjerns lets leave this as `0.3.0.dev0`

skjerns replied:
done

dalthviz replied:
Seems unchanged here, maybe a missing file save before doing the commit @skjerns ?

skjerns replied:
Ah, yes, commited from the wrong folder, thanks!

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+    CONF_WIDGET_CLASS = SpyderLineProfilerConfigPage
+
+    # --- Signals
+    # sig_finished = Signal()
```
Probably we should keep this signal, right?

skjerns replied:
done

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+
+    def on_initialize(self):
+        self.widget = self.get_widget()
+        # self.widget.sig_finished.connect(self.sig_finished)
```
Probably this should be uncommented? Did you face some issue when connecting the signal from the widget to the plugin?

skjerns replied:
I found not difference at all when disconnecting the signal. Tbh. I don't understand where the signal comes into play at all, or at which point it would be necessary. Currently the `finished` method is connected to the subprocess exiting.

I'd appreciate if you have a pointer in the documentation to read about what the signals do and in which case they are emitted by Spyder, I could not find it. Internally in the line profiler plugin, I don't think it's actually used anymore.

skjerns replied:
So I can either 
- remove the finished signals altogether, as I don't see them being connected to anything meaningful, or
-  if they have some function within Spyder that I'm not aware of I'll put them back

dalthviz replied:
I think the idea of the signal is facilitate other plugins to be aware of the line profiler finishing execution, so kind of as an extention point/way to comunicate with other plugins. However, as you say, seems like there is no much usage for it at the moment. What do you think @ccordoba12 and @jitseniesen, should we keep it or remove it?

dalthviz replied:
Checking this @skjerns lets leave the signal. Although is not used currently, you could think of it as part of the ways the plugin has to comunicate its state to not only Spyder but other plugins so it could be useful in the future.

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -68,7 +58,29 @@
+                                   adjust_to_contents=adjust_to_contents)
+        except Exception:
+            PythonModulesComboBox.__init__(self, parent=parent,
+                                       adjust_to_contents=adjust_to_contents)
```
I think that with the changes in the CI workflows configs this shouldn't be needed but lets check first applying the config🤞🏼

skjerns replied:
Ok, I removed the patch

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +112,217 @@ def natural_sort_key(key):
+            SpyderLineProfilerWidgetActions.Collapse,
+            text=_("Collaps"),
+            tip=_('Collapse all'),
+            icon=qta.icon("mdi.arrow-collapse-all", color=ima.MAIN_FG_COLOR),
```
This should be:

```suggestion
            icon=self.create_icon('collapse'),
```
Which will make it look like this:
![imagen](https://user-images.githubusercontent.com/16781833/159550418-763feaf1-6605-4da0-a76b-644982c45174.png)

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +112,217 @@ def natural_sort_key(key):
-        self.collapse_button = create_toolbutton(
-            self,
-            icon=get_icon('collapse.png'),
+            icon=qta.icon("fa.file-text", color=ima.MAIN_FG_COLOR),
```
I checked locally and this should be:

```suggestion
            icon=self.create_icon('log'),
```
Which will make it look like this:
![imagen](https://user-images.githubusercontent.com/16781833/159551013-ea3f1d12-4cbf-4e88-b2c7-9c9f9989d93c.png)

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +112,217 @@ def natural_sort_key(key):
+            SpyderLineProfilerWidgetActions.Expand,
+            text=_("Expand"),
+            tip=_('Expand all'),
+            icon=qta.icon("mdi.arrow-expand-all", color=ima.MAIN_FG_COLOR),
```
This should be:
```suggestion
            icon=self.create_icon('expand'),
```
Which will make it look like this:
![imagen](https://user-images.githubusercontent.com/16781833/159551439-8213fbcf-8cee-4963-8293-59f1e53ef0c1.png)

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +112,217 @@ def natural_sort_key(key):
+            SpyderLineProfilerWidgetActions.Clear,
+            text=_("Clear output"),
+            tip=_('Clear'),
+            icon=qta.icon("fa.trash", color=ima.MAIN_FG_COLOR),
```
This should be:
```suggestion
            icon=self.create_icon('editdelete'),
```
Which will make it look like this:
![imagen](https://user-images.githubusercontent.com/16781833/159552156-7d31d744-d4e3-46c9-9144-d1653a12e79d.png)

dalthviz reviewed (commented):
Thank you so much for the patience with this @skjerns I left some more suggestions regarding the icon for the actions and also the CI config to hopefully prevent the need to monkeypatch stuff 🤞🏼 

Just in case, with the suggested icon changes this is how is looking on my local setup:

![imagen](https://user-images.githubusercontent.com/16781833/159553062-af55d71a-b48e-44b1-9344-32dc2c6204a9.png)

skjerns commented:
Thanks for all the changes! I think it's ready now :)

Btw. I think the plugin how-to needs to be update w.r.t the icons, it is currently still mentioning having a look at the `qta-browser`. Might be good to have a way to see which icons are available builtin: https://docs.spyder-ide.org/current/workshops/plugin-development.html

dalthviz commented:
Thanks @skjerns for the quick response to the review ! Seems like this is getting closer and closer to be ready 🎉 :+1:

Regarding the missing things:

> Now we need to decide to ditch older Python versions (as they are incompatible with Spyder5, even though it seems to run fine when installing with pip, but the requirement spyer>=5.2 makes them not able to run)

I would say that we should remove then the support for Python <=3.6, but just in case, what dou you think @ccordoba12 @jitseniesen ?

> and I need to know if the signals are strictly necessary (see above).

I left a new comment in the review with a possible explanation but probably having someone else opinion could helps us decide if is worthy or not to keep it (as you say seems like there is no use for it currently but maybe there is some idea behind it).

>Btw. I think the plugin how-to needs to be update w.r.t the icons, it is currently still mentioning having a look at the qta-browser. Might be good to have a way to see which icons are available builtin by creating a spyder-qta-browser: https://docs.spyder-ide.org/current/workshops/plugin-development.html

I think that is a quite good idea, could you open an issue for that in the Spyder repo to keep track of it? Thanks!

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+        mainmenu = self.get_plugin(Plugins.MainMenu)
+
+        mainmenu.remove_item_from_application_menu(
+            SpyderLineProfilerActions.ProfileCurrentFile,
```
This should be `SpyderLineProfilerActions.Run`
```suggestion
            SpyderLineProfilerActions.Run,
```

dalthviz replied:
Checked again and `SpyderLineProfilerActions` doesn't have a `ProfileCurrentFile` so my guess is that my old suggestion should be used

skjerns commented:
I realize that the with the current version (spyder 5.3) and changes the pane appears at its own pane and not as a tab within the the other panes. do you know why?

![image](https://user-images.githubusercontent.com/14980558/161700073-23e98f0e-d280-4940-9440-6bfd37c18241.png)

skjerns commented:
I'm not quite sure what is happening here, as I changed several things (reinstalled my environment with miniconda which comes with Py39 instead of Py38, updated Spyder to 5.3, lot's of changes in this repo as well)

So I'm not sure which change exactly caused the problem as I cannot reproduce the previous state, even when resetting everything (except exchanging miniconda with anaconda) to before. I'll investigate

skjerns commented:
> I would say that we should remove then the support for Python <=3.6, but just in case, what dou you think @ccordoba12 @jitseniesen ?

As Spyder 5.3 removes support for Python 3.6 as well, I think we can go ahead :)

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -257,17 +345,25 @@ def analyze(self, filename=None, wdir=None, args=None, pythonpath=None,
         if self.output:
-            editor = TextEditor(self.output, title=_("Line profiler output"),
-                                readonly=True)
+            editor = TextEditor(self.output, title=_("Line profiler output"))
```
I think we are missing the `parent=self` kwarg when creating the `TextEditor` instance since right now is showing as if Spyder was using the clear theme:

![imagen](https://user-images.githubusercontent.com/16781833/162042520-43e11ddc-e98f-4bd5-a5b9-fe75f558f61d.png)

We also should add it in the `TextEditor` creation at the `show_errorlog` dialog

Also, I think `readonly=True` kwarg should still be passed since the `save and close` button doesn't do anything (as far as I tested) or maybe is there some reason why the button should be shown?

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -424,7 +580,7 @@ def __init__(self, parent=None):
         self.setHeaderLabels(self.header_list)
         self.clear()
-        self.itemActivated.connect(self.item_activated)
+        self.itemClicked.connect(self.onItemClick)
```
Maybe this function should be called `on_item_clicked`?
```suggestion
        self.itemClicked.connect(self.on_item_clicked)
```
Also, is there some specific reason why `itemClicked` is better than `itemActivated`?

skjerns replied:
Nope, I simply copied this piece of coded from somewhere else, so no reason.

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -499,7 +655,8 @@ def load_data(self, profdatafile):
             self.stats[func_info] = [func_stats, func_total_time]
-
+            
+       
```
Extra blank line that can be removed

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -613,23 +770,74 @@ def populate_tree(self):
                 # Monospace font for code
                 line_item.setFont(COL_LINE, monospace_font)
+                
+    def onItemClick(self, item):
```
Following the suggestion above:
```suggestion
    def on_item_clicked(self, item):
```

dalthviz reviewed (commented):
Thanks again @skjerns for all the progress done with this one. I gave it another pass and added some suggestions regarding the widget to show the output and the name of a method. 

Besides that I was thinking and maybe we will need to update the README with a new screenshot of the plugin and also update the description there to add that the plugin is compatible with Spyder > 5.2 (even maybe we could create a table to show the compatibility relationship between the plugin versions and the Spyder versions). Would you like to do the README update here or maybe leave it for a future PR?

Let us know and thanks again for all the hard work!

skjerns commented:
Included the suggestions.

Now the window is finally inheriting the spyder theme! great :)

ccordoba12 commented:
Thanks @skjerns for all your hard work on this! I'll try to leave my review in a couple of days, but it seems @dalthviz and you have done a great work so far.

ccordoba12 commented on spyder_line_profiler/spyder/api.py:
```diff
@@ -0,0 +1,9 @@
+# -*- coding: utf-8 -*-
```
I think this file is not necessary because this plugin doesn't expose any API.

skjerns replied:
ok removed

ccordoba12 commented on spyder_line_profiler/spyder/confpage.py:
```diff
@@ -0,0 +1,58 @@
+# ----------------------------------------------------------------------------
+"""
+Spyder Line Profiler 5 Preferences Page.
+"""
```
```suggestion
# ----------------------------------------------------------------------------

"""
Spyder Line Profiler 5 Preferences Page.
"""

```

Add blank lines for clarity.

ccordoba12 commented on spyder_line_profiler/spyder/confpage.py:
```diff
@@ -0,0 +1,58 @@
+from spyder.api.preferences import PluginConfigPage
+from spyder.api.translations import get_translation
+from .widgets import SpyderLineProfilerWidget
+_ = get_translation("spyder_line_profiler.spyder")
```
```suggestion
from spyder.api.preferences import PluginConfigPage
from spyder.api.translations import get_translation

# Local imports
from .widgets import SpyderLineProfilerWidget

# Localization
_ = get_translation("spyder_line_profiler.spyder")
```

* Spyder imports are third-party for this plugin
* The only local import is SpyderLineProfilerWidget
* The localization statement shouldn't be mixed with the imports.

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+# ----------------------------------------------------------------------------
+"""
+Spyder Line Profiler 5 Plugin.
+"""
```
```suggestion
# ----------------------------------------------------------------------------

"""
Spyder Line Profiler 5 Plugin.
"""
```

For clarity

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+from spyder_line_profiler.spyder.widgets import SpyderLineProfilerWidgetActions
+from spyder_line_profiler.spyder.widgets import is_lineprofiler_installed
+
+_ = get_translation("spyder_line_profiler.spyder")
```
```suggestion
# Third-party imports
from qtpy.QtCore import Signal
from qtpy.QtGui import QIcon
from spyder.api.plugins import Plugins, SpyderDockablePlugin
from spyder.api.translations import get_translation
from spyder.api.plugin_registration.decorators import on_plugin_available
from spyder.api.plugin_registration.decorators import on_plugin_teardown
from spyder.plugins.mainmenu.api import ApplicationMenus

# Local imports
from spyder_line_profiler.spyder.confpage import SpyderLineProfilerConfigPage
from spyder_line_profiler.spyder.widgets import SpyderLineProfilerWidget
from spyder_line_profiler.spyder.widgets import SpyderLineProfilerWidgetActions
from spyder_line_profiler.spyder.widgets import is_lineprofiler_installed

# Localization
_ = get_translation("spyder_line_profiler.spyder")
```

Rearrange imports.

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+
+class SpyderLineProfilerActions:
+    # Triggers
+    Run = 'run_action_menu'
```
```suggestion
    Run = 'run_profiler_action'
```

I think this is more telling.

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+        return _("Line Profiler")
+
+    def get_description(self):
+        return _("Line profiler display for Spyder 5")
```
```suggestion
        return _("Line profiler display for Spyder")
```

I think it's not necessary to mention the version.

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+                        icon=self.get_icon(),
+                        triggered=self.run_lineprofiler,
+                        register_shortcut=True,
+                        )   
```
```suggestion
        run_action = self.create_action(
            SpyderLineProfilerActions.Run,
            text=_("Run line profiler"),
            tip=_("Run line profiler"),
            icon=self.get_icon(),
            triggered=self.run_lineprofiler,
            register_shortcut=True,
        )   
```

Fix indentation.

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+                        )   
+        run_action.setEnabled(is_lineprofiler_installed())
+        
+        
```
```suggestion
```

Remove unnecessary blank.

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+        mainmenu = self.get_plugin(Plugins.MainMenu)
+        start_action = self.get_action(SpyderLineProfilerActions.Run)
+        mainmenu.add_item_to_application_menu(
+            start_action, menu_id=ApplicationMenus.Run)
```
```suggestion
        run_action = self.get_action(SpyderLineProfilerActions.Run)
        mainmenu.add_item_to_application_menu(
            run_action, menu_id=ApplicationMenus.Run)
```

Use a more telling name for this variable.

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+    @on_plugin_teardown(plugin=Plugins.MainMenu)
+    def on_main_menu_teardown(self):
+        mainmenu = self.get_plugin(Plugins.MainMenu)
+
```
```suggestion
```

Remove unnecessary blank.

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+            menu_id=ApplicationMenus.Run
+        )    
+        
+        
```
```suggestion
```

Unnecessary blank.

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+    def on_close(self, cancellable=True):
+        return True
+        
+
```
```suggestion
```

Unnecessary blank.

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+        """
+        self.widget.spyder_pythonpath = self.main.get_spyder_pythonpath()
+        
+        
```
```suggestion
```

Unnecessary blank.

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+        self.analyze(self.main.editor.get_current_filename())
+
+    def analyze(self, filename):
+        """Reimplement analyze method."""        
```
```suggestion
        """Analyze a file."""
```

Fix docstring.

ccordoba12 commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,133 @@
+    """
+
+    NAME = "spyder_line_profiler"
+    REQUIRES = [Plugins.Editor, Plugins.Help]
```
```suggestion
    REQUIRES = [Plugins.Editor]
```

* This field is is to require other plugins without which this one is unable to work. Since Help is not needed like that, I think we don't need to add it here.
* The Layout plugin will take care of tabifying this plugin correctly even if Help is not available.

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -1,14 +1,11 @@
-
-See the official documentation of line_profiler:
-http://pythonhosted.org/line_profiler/
+Spyder Line Profiler 5 Main Widget.
```
```suggestion
Spyder Line Profiler Main Widget.
```

No need to mention the version.

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -68,7 +58,17 @@
+
+def is_lineprofiler_installed():
+    """
+    Checks if the program and the library for line_profiler is installed.
```
```suggestion
    Check if the program and the library for line_profiler is installed.
```

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,210 @@ def natural_sort_key(key):
+    SaveData = 'save_data_action'
+    ShowOutput = 'show_output_action'
+    Stop = 'stop_action'
+    
```
```suggestion


```

Double space among class declarations

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,210 @@ def natural_sort_key(key):
+    Main = 'main_section'
+    ExpandCollaps = 'expand_collaps_section'
+    ShowOutput = 'show_output_section'
+    
```
```suggestion


```

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,210 @@ def natural_sort_key(key):
+    
+class SpyderLineProfilerWidgetToolbars:
+    Information = 'information_toolbar'
+    
```
```suggestion


```

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,210 @@ def natural_sort_key(key):
+    
+class SpyderLineProfilerWidgetMainToolbarItems:
+    FileCombo = 'file_combo'
+    
```
```suggestion


```

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,210 @@ def natural_sort_key(key):
+    
+class SpyderLineProfilerWidgetInformationToolbarSections:
+    Main = 'main_section'
+    
```
```suggestion


```

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,210 @@ def natural_sort_key(key):
+    Stretcher1 = 'stretcher_1'
+    Stretcher2 = 'stretcher_2'
+    DateLabel = 'date_label'  
+  
```
```suggestion


```

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,210 @@ def natural_sort_key(key):
+        self.filecombo = PythonModulesComboBox(
+            self, id_= SpyderLineProfilerWidgetMainToolbarItems.FileCombo)
+        self.datatree = LineProfilerDataTree(self)
+        self.datelabel = QLabel()
```
```suggestion
        self.datelabel = QLabel(self)
```

I think this should have `self` as parent to avoid segfaults when closing Spyder, right @dalthviz?

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,210 @@ def natural_sort_key(key):
+        self.datelabel.ID = SpyderLineProfilerWidgetInformationToolbarItems.DateLabel
+        self.datelabel.setText(_('Please select a file to profile, with '
+                                 'added @profile decorators for functions'))
+        self.timer = QTimer()
```
```suggestion
        self.timer = QTimer(self)
```

And the same for this one.

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,210 @@ def natural_sort_key(key):
+        
+    # --- PluginMainWidget API
+    # ------------------------------------------------------------------------
+    
```
```suggestion
```

Unnecessary blank.

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,210 @@ def natural_sort_key(key):
+                toolbar=toolbar,
+                section=SpyderLineProfilerWidgetMainToolbarSections.Main,
+            )
+        # Secondary Toolbar
```
```suggestion

        # Secondary Toolbar
```

For clarity.

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,210 @@ def natural_sort_key(key):
+                section=SpyderLineProfilerWidgetInformationToolbarSections.Main,
+            )
+
+        
```
```suggestion
```

Unnecessary blanks.

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +479,97 @@ def read_output(self, error=False):
-        # FIXME: figure out if show_data should be called here or
-        #        as a signal from the combobox
-        self.show_data(justanalyzed=True)
+        if not self.output=='aborted':
```
```suggestion
        if not self.output == 'aborted':
```

Missing spaces.

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +479,97 @@ def read_output(self, error=False):
+                f.write(results)
+                
+            self.datelabel.setText(_(f"Saved results to {filename}"))
+        
```
```suggestion
```

Unnecessary blank.

ccordoba12 commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +479,97 @@ def read_output(self, error=False):
+    def on_section_conf_change(self, section):
+        pass
+
+
```
```suggestion
```

Unnecessary blanks.

ccordoba12 commented on spyder_line_profiler/tests/test_lineprofiler.py:
```diff
@@ -17,7 +17,11 @@
+import spyder
+import spyder_line_profiler
+from spyder_line_profiler.spyder.widgets import SpyderLineProfilerWidget
+from spyder_line_profiler.spyder.widgets import PythonModulesComboBox
```
```suggestion
from spyder_line_profiler.spyder.widgets import SpyderLineProfilerWidget
```

These three imports don't seem necessary.

ccordoba12 commented on spyder_line_profiler/tests/test_lineprofiler.py:
```diff
@@ -34,6 +38,8 @@ def foo():
     for k in range(100):
         xs = xs + ['x']
 foo()"""
+
```
```suggestion
```

Unnecessary blank.

skjerns commented:
@ccordoba12 thanks, that was a thorough review! sorry for all the missing or additional spaces, I should probably consider using an auto-formatter or similar :-)

dalthviz commented:
Thanks @ccordoba12 for the review! 

@skjerns to accept all the suggestions at once you can use the `Add suggestion to batch` functionality (although I think you need to be in the `Files changed` view): https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes

dalthviz commented:
@skjerns could it possible for you to squash the commits related to suggestions into single commits? We where discussing with @ccordoba12 and that could be nice to have a clean commit history here. If you need help with that let us know.

After that I will test again locally and hopefully approve and merge this. Also, just in case, pinging @jitseniesen to check if that sounds good and maybe to get more feedback

jitseniesen commented:
> Also, just in case, pinging @jitseniesen to check if that sounds good and maybe to get more feedback

@dalthviz I did not look at it in detail, but I did have a quick look a couple of times and it all looks reasonable. I think this PR has had enough scrutiny so I'm happy for you to go ahead.

skjerns commented:
@dalthviz not sure if the squash/rebase worked? let me know if this is what you wanted, else I would need to know how to perform this.

dalthviz commented:
Seems like the commits are still there :/

I think now the process is a little bit more involved to do due to the merge commits. What I thinkcould work is trying to use `git reflog` and `git reset --hard` to get to the previous branch state (before the last merge commits) and then use `git rebase -i HEAD~31` to squash/fixup the 31 commits that where done to address suggestions

If is ok for you @skjerns I could try to do that and then force push the changes here. Let me know if you want to try or if I should go ahead

skjerns commented:
So, I think that did the trick! Thanks, never dared to work with rebase so far.

dalthviz commented:
Awesome! And yes rebasing and force pushing are kind of terrifying at first but once you get the grip of it (and know a way to undo them if needed) are quite useful :)

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+# Local imports
+from spyder_line_profiler.spyder.confpage import SpyderLineProfilerConfigPage
+from spyder_line_profiler.spyder.widgets import SpyderLineProfilerWidget
+from spyder_line_profiler.spyder.widgets import SpyderLineProfilerWidgetActions
```
Seems like this is an unsued import
```suggestion

```

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+        return _("Line profiler display for Spyder")
+
+    def get_icon(self):
+        return QIcon('./data/images/spyder.line_profiler.png')
```
Checking seems like this is not getting properly resolved to the actual image. We can change it for something like this:
```suggestion
        path = osp.join(
            osp.dirname(osp.dirname(__file__)),
            'data', 'images', 'spyder.line_profiler.png')
        return QIcon(path)
```

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+"""
+Spyder Line Profiler 5 Plugin.
+"""
+
```
Adds the needed import for the `get_icon` path construction logic:
```suggestion

# Standard library imports
import os.path as osp

```

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+    """
+
+    NAME = "spyder_line_profiler"
+    REQUIRES = [Plugins.Editor]
```
We are missing the preferences plugin here:
```suggestion
    REQUIRES = [Plugins.Preferences, Plugins.Editor]
```

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+    TABIFY = [Plugins.Help]
+    WIDGET_CLASS = SpyderLineProfilerWidget
+    CONF_SECTION = NAME
+    CONF_WIDGET_CLASS = SpyderLineProfilerConfigPage
```
Missing `CONF_FILE` set to `True`
```suggestion
    CONF_WIDGET_CLASS = SpyderLineProfilerConfigPage
    CONF_FILE = True
```

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+            tip=_("Run line profiler"),
+            icon=self.get_icon(),
+            triggered=self.run_lineprofiler,
+            register_shortcut=True,
```
Missing context attribute to be able to use the shortcut from the application context:
```suggestion
            context=Qt.ApplicationShortcut,
            register_shortcut=True,
```

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+"""
+
+# Third-party imports
+from qtpy.QtCore import Signal
```
To add `Qt` import needed for the new context argument for the run action:
```suggestion
from qtpy.QtCore import Qt, Signal
```

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+            SpyderLineProfilerActions.ProfileCurrentFile,
+            menu_id=ApplicationMenus.Run
+        )    
+        
```
Missing plugin Preferences registration:
```suggestion

    @on_plugin_available(plugin=Plugins.Preferences)
    def on_preferences_available(self):
        preferences = self.get_plugin(Plugins.Preferences)
        preferences.register_plugin_preferences(self)

```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -19,41 +16,34 @@
-from qtpy.compat import getopenfilename
-from qtpy.QtCore import (QByteArray, QProcess, Qt, QTextCodec,
-                         QProcessEnvironment, Signal)
+from qtpy.QtWidgets import QHBoxLayout, QLabel
```
Seems like `QHBoxLayout` is an unused import and in a later line `qtpy.QtWidgets` elements are being import so we could delete this line and add `QLabel` to the import statement above:
```suggestion

```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -19,41 +16,34 @@
-                            QLabel, QTreeWidget, QTreeWidgetItem, QApplication)
+from qtpy.QtCore import (QByteArray, QProcess, Qt, QTextCodec,
+                         QProcessEnvironment, Signal, QTimer)
+from qtpy.QtWidgets import (QMessageBox, QVBoxLayout,
```
`QLabel` import from the removed line:
```suggestion
from qtpy.QtWidgets import (QMessageBox, QVBoxLayout, QLabel,
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -19,41 +16,34 @@
+from qtpy.QtWidgets import (QMessageBox, QVBoxLayout,
+                            QTreeWidget, QTreeWidgetItem, QApplication)
+from qtpy.compat import getopenfilename, getsavefilename
+import qtawesome as qta
```
Seems like an unused import:
```suggestion
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -19,41 +16,34 @@
+from spyder.utils.misc import add_pathlist_to_PYTHONPATH, getcwd_or_home
+from spyder.plugins.run.widgets import get_run_configuration
+from spyder.py3compat import to_text_string, getcwd, pickle
+from spyder.utils.icon_manager import ima
```
Seems like `getcwd` and `ima` imports are not being used:
```suggestion
from spyder.py3compat import to_text_string, pickle
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -1,14 +1,11 @@
+Spyder Line Profiler Main Widget.
 """
 # Standard library imports
 from __future__ import with_statement
```
I think we can remove this import too:
```suggestion
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -68,7 +58,17 @@
+    return (programs.is_module_installed('line_profiler')
+            and programs.find_program('kernprof') is not None)
+
+    
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
+class SpyderLineProfilerWidgetInformationToolbarItems:
+    Stretcher1 = 'stretcher_1'
+    Stretcher2 = 'stretcher_2'
+    DateLabel = 'date_label'  
```
Remove trailing whitespace:
```suggestion
    DateLabel = 'date_label'
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
+    """
+    This signal will request to open a file in a given row and column
+    using a code editor.
+    
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
+        self.spyder_pythonpath = None
+        self.process = None
+        self.started_time = None
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
+        
+        # Widgets
+        self.filecombo = PythonModulesComboBox(
+            self, id_= SpyderLineProfilerWidgetMainToolbarItems.FileCombo)
```
Remove space around parameter equals:
```suggestion
            self, id_=SpyderLineProfilerWidgetMainToolbarItems.FileCombo)
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
+        layout = QVBoxLayout()
+        layout.addWidget(self.datatree)
+        self.setLayout(layout)
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
+        # Signals
+        self.datatree.sig_edit_goto_requested.connect(
+            self.sig_edit_goto_requested)
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
+        pass
+
+    def setup(self):
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
+        self.browse_action = self.create_action(
+            SpyderLineProfilerWidgetActions.Browse,
+            text=_("Open Script"),
+            tip=_('Select Python script'),            
```
Remove trailing whitespace:
```suggestion
            tip=_('Select Python script'),
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
+        self.clear_action.setEnabled(False)
+        self.log_action.setEnabled(False)
+        self.save_action.setEnabled(False)
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
+        
+        # Main Toolbar
+        toolbar = self.get_main_toolbar()
+        for item in [self.filecombo, self.browse_action, self.start_action, 
```
Remove trailing whitespace:
```suggestion
        for item in [self.filecombo, self.browse_action, self.start_action,
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
+                toolbar=secondary_toolbar,
+                section=SpyderLineProfilerWidgetInformationToolbarSections.Main,
+            )
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
             self.datelabel.setOpenExternalLinks(True)
         else:
-            pass  # self.show_data()
+            pass  
```
Remove trailing whitespace:
```suggestion
            pass
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -100,123 +100,214 @@ def natural_sort_key(key):
-            pass  # self.show_data()
+            pass  
 
+        
```
Remove blank line:
```suggestion
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -233,9 +324,10 @@ def analyze(self, filename=None, wdir=None, args=None, pythonpath=None,
             else:
                 self.filecombo.setCurrentIndex(index)
             self.filecombo.selected()
+            
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -257,17 +349,26 @@ def analyze(self, filename=None, wdir=None, args=None, pythonpath=None,
+        except:
+            curr_script = None
+            
+        pwd = curr_script if curr_script else getcwd_or_home()
```
`self.editor` is not defined so I guess this always goes to the expect path. If that is the case, this could be changed to:
```suggestion
        pwd = getcwd_or_home()
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -257,17 +349,26 @@ def analyze(self, filename=None, wdir=None, args=None, pythonpath=None,
+            curr_script = None
+            
+        pwd = curr_script if curr_script else getcwd_or_home()
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -257,17 +349,26 @@ def analyze(self, filename=None, wdir=None, args=None, pythonpath=None,
+            self, _("Select Python script"), pwd,
             _("Python scripts")+" (*.py ; *.pyw)")
         self.redirect_stdio.emit(False)
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -257,17 +349,26 @@ def analyze(self, filename=None, wdir=None, args=None, pythonpath=None,
         if self.output:
             editor = TextEditor(self.output, title=_("Line profiler output"),
-                                readonly=True)
+                                                    readonly=True, parent=self)
```
Fix indentation:
```suggestion
                                readonly=True, parent=self)
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -257,17 +349,26 @@ def analyze(self, filename=None, wdir=None, args=None, pythonpath=None,
             editor = TextEditor(self.output, title=_("Line profiler output"),
-                                readonly=True)
+                                                    readonly=True, parent=self)
+            
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -277,38 +378,46 @@ def show_errorlog(self):
             # see spyder-ide/spyder#12202
             editor.show()
             editor.exec_()
+    
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -277,38 +378,46 @@ def show_errorlog(self):
                 args = []
         if pythonpath is None:
             pythonpath = self._last_pythonpath
+            
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -277,38 +378,46 @@ def show_errorlog(self):
-
+        self.datelabel.setText(_('Profiling starting up, please wait...'))
+        self.started_time = datetime.now()
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
+        self.show_errorlog()  # If errors occurred, show them.
         self.sig_finished.emit()
-
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
+                self.output = 'aborted'
                 self.process.waitForFinished()
-
+                
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
+        self.log_action.setEnabled(False)
+        self.save_action.setEnabled(False)
+        self.output = ''
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
+        self.clear_action.setEnabled(output_exists)
+        self.log_action.setEnabled(output_exists)
+        self.save_action.setEnabled(output_exists)
+
```
Remove whitespace on blank line

skjerns replied:
I think there is none

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
         date_text = text_style % time.strftime("%d %b %Y %H:%M",
                                                time.localtime())
         self.datelabel.setText(date_text)
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
+        if not self.output:
+            self.datelabel.setText(_("Nothing to save"))
+            return
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
+            self.datelabel.setText(_("Nothing to save"))
+            return
+        
+        title = _( "Save line profiler result")
```
Remove whitespace:
```suggestion
        title = _("Save line profiler result")
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
+        filename, _selfilter = getsavefilename(
+            self,
+            title,
+             f'{curr_filename}_lineprof.txt',
```
Fix indentation:
```suggestion
            f'{curr_filename}_lineprof.txt',
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
+            with open(filename, 'w') as f:
+                # for some weird reason, everything is double spaced on Win
+                results = self.output
+                print(results)
+                results  = results.replace('\r', '')
+                print(results.encode())
```
Remove `print` statements and extra space:
```suggestion
                results = results.replace('\r', '')
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
+                results  = results.replace('\r', '')
+                print(results.encode())
+                f.write(results)
+                
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
+                f.write(results)
+                
+            self.datelabel.setText(_(f"Saved results to {filename}"))
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
+    @on_conf_change
+    def on_section_conf_change(self, section):
+        pass
+    
+    
```
Remove whitespace on blank lines

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -371,48 +483,94 @@ def read_output(self, error=False):
-    edit_goto = Signal(str, int, str)
-
+    sig_edit_goto_requested = Signal(str, int, str)
+    
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -613,23 +771,74 @@ def populate_tree(self):
 
                 # Monospace font for code
                 line_item.setFont(COL_LINE, monospace_font)
+                
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -613,23 +771,74 @@ def populate_tree(self):
+                
+    def on_item_clicked(self, item):
+        data = item.data(COL_POS, Qt.UserRole)
+        if data is None or len(data)<2: return
```
Missing spaces and `return` in a new line:
```suggestion
        if data is None or len(data) < 2:
            return
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -613,23 +771,74 @@ def populate_tree(self):
+# Tests
+# =============================================================================
+
+profile = lambda  x:x # dummy profile wrapper to make script load externally
```
Some issues with spaces:
```suggestion
profile = lambda x: x  # dummy profile wrapper to make script load externally
```

dalthviz commented on spyder_line_profiler/spyder/widgets.py:
```diff
@@ -613,23 +771,74 @@ def populate_tree(self):
+# =============================================================================
+
+profile = lambda  x:x # dummy profile wrapper to make script load externally
+
```
Add extra blank line:
```suggestion


```

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+_ = get_translation("spyder_line_profiler.spyder")
+
+
+
```
Remove one blank line:
```suggestion
```

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+class SpyderLineProfilerActions:
+    # Triggers
+    Run = 'run_profiler_action'
+    
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+    def on_initialize(self):
+        self.widget = self.get_widget()
+        self.widget.sig_finished.connect(self.sig_finished)
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+            icon=self.get_icon(),
+            triggered=self.run_lineprofiler,
+            register_shortcut=True,
+        )   
```
Remove trailing whitespace:
```suggestion
        )
```

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+            register_shortcut=True,
+        )   
+        run_action.setEnabled(is_lineprofiler_installed())
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+        run_action = self.get_action(SpyderLineProfilerActions.Run)
+        mainmenu.add_item_to_application_menu(
+            run_action, menu_id=ApplicationMenus.Run)
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+        mainmenu.remove_item_from_application_menu(
+            SpyderLineProfilerActions.ProfileCurrentFile,
+            menu_id=ApplicationMenus.Run
+        )    
```
Remove trailing whitespace:
```suggestion
        )
```

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+
+    def on_close(self, cancellable=True):
+        return True
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+        PYTHONPATH in Spyder.
+        """
+        self.widget.spyder_pythonpath = self.main.get_spyder_pythonpath()
+        
```
Remove whitespace on blank line

dalthviz commented on spyder_line_profiler/spyder/plugin.py:
```diff
@@ -0,0 +1,128 @@
+        if editor.save():
+            self.switch_to_plugin()
+            self.analyze(editor.get_current_filename())
+        
```
Remove whitespace on blank line

dalthviz reviewed (commented):
Hi again @skjerns ! I did a full check locally and added some suggestions accordingly. Mostly all the changes are to improve the code style (remove whitespaces on blank lines or trailing whitespaces). Besides that, I realized some shortcuts functionality config and preferences registration where missing so I added the respective code as suggestions. If you have any questions or comments let me know!

skjerns commented:
whoops, that broke something on Linux. Unfortunately the error message is rather inconclusive and I'm on Windows, so can't really debug this. Any idea how to fix this?

dalthviz commented:
Thanks for all the work! I will do a check for the Linux tests :+1: (maybe the errors are related with PyQT 5.15 being no available on conda-forge)

dalthviz commented on spyder_line_profiler/tests/test_lineprofiler.py:
```diff
@@ -17,7 +17,8 @@
 MAIN_APP = qapplication() 
```
Could you remove this line and the import related with `qapplication`? I think that is what is causing the segfault on the Linux CI

dalthviz commented on spyder_line_profiler/tests/test_lineprofiler.py:
```diff
@@ -17,7 +17,8 @@
+
 
 try:
     from unittest.mock import Mock
```
Checking this file maybe is worthy also to remove this try-except import logic since it is doing just a fallback for Python 2

dalthviz commented:
So seems like that made the error raise somewhere else 🤔  I will keep trying other changes and see if that helps. If is okay with you then I would push them directly here.

Thank you so much for all the effort and time in this! Hopefully a new release with all these changes will be available soon™ !

dalthviz reviewed (approved):
Now tests are passing so is time to merge! Thanks @skjerns !

ccordoba12 commented:
Thanks a lot @skjerns for your help to update this plugin!

skjerns commented:
🥳🥳🥳🥳

thanks guys!

nosterlu commented:
Thank you so much, really looking forward to using this again in Spyder!
